org.lflang/src/org/lflang/generator/c/CCmakeGenerator.java

```diff
@@ -69,6 +69,16 @@
      */
     StringBuilder generateCMakeCode(List<String> sources, String executableName, ErrorReporter errorReporter) {
         StringBuilder cMakeCode = new StringBuilder();
+        // Decide if the compilation should happen in C++ mode
+        boolean CppMode = false;
+        if (targetConfig.compiler.equals("g++") || 
+                targetConfig.compiler.equals("CC") ||
```

Comment:
Case?

Reply:
It is case sensitive I think. `CC` is the C++ compiler and `cc` is the C compiler on macOS apparently.

Reply:
This looks a bit fishy to me. Which compiler is `CC`? I now `CC` only as a environment variable that indicates the C compiler to use for compilation.

I don't think that pattern matching is a reliable way of finding the user's intent here. Wouldn't it be better to have an explicit target property `c++-mode` or something similar?

Reply:
This looks a bit fishy to me. Which compiler is `CC`? I now `CC` only as an environment variable that indicates the C compiler to use for compilation.

I don't think that pattern matching is a reliable way of finding the user's intent here. Wouldn't it be better to have an explicit target property c++-mode or something similar?

Reply:
It is a bit of a workaround, and I don't think it is a permanent solution. `cc` and `CC` can both be present on a system as a symlink. It is hard to nail down a convention, but I think `cc` is used for the C compiler and `CC` is used for the C++ compiler.

The reason I didn't add a `c++-mode` is because, first, we have a Cpp target. If someone wants C++ functionality, then they should definitely use the Cpp target. Second, it can be misleading as to what C++ mode does. Does it turn reactors into C++ classes? Does it make the runtime more memory-safe? Currently, it would only switch the compiler to a C++ compiler. So a more accurate flag would be `use-cpp-compiler` perhaps.

Reply:
@Soroush: I think that your explanation underlines that the situation is too complex for having a bunch of hidden assumptions. It's not at all obvious what is the right thing to do here. I, too, wondered what intent this code was intended to capture. Explicit indicators would be appropriate. I think a target property `use-cpp-compiler` makes sense, but we already have a target property called `compiler`, which that seems to overlap with.

Reply:
I think the underlying issue that started this set of hidden triggers that enables a `CppMode` was the fact that CMake does not like to apply a C++ compiler to `.c` files. Without the `CMake` build system (using the legacy build system), setting the compiler to `g++` would work fine for all the scenarios that I was concerned with. With CMake, [this](https://github.com/icyphy/lingua-franca/blob/b71841e3436825cb0a06e51497a003c813023410/org.lflang/src/org/lflang/generator/c/CCmakeGenerator.java#L173) line needs to be added to the CMakeLists.txt so that the `.c` files are interpreted as C++ files.

Reply:
I also remembered that an [exception](https://github.com/icyphy/lingua-franca/blob/e9c9bcb70e775d08edbf300668a740065a20b1c1/org.lflang/src/org/lflang/generator/c/CCmakeCompiler.java#L83) has to be made in setting the environment variable that tells CMake what compiler to use to exclude these C++ compilers. Otherwise, the same error will be thrown even after adding the aforementioned line:

    error: "The CMAKE_C_COMPILER is set to a C++ compiler"

Reply:
Ultimately, if an unknown C++ compiler is requested by the user, CMake will throw an error at compile time. I've also made changes so that the known C++ compilers (g++, CC, and clang++) are actually applied as requested by the user in the `compiler` target property (previously, they were ignored in favor of a CMake-detected C++ compiler).

Reply:
There seem to be two issues entangled here:

1. a directive to use a specific compiler (the `compiler` target property)
2. a directive to use a kind of compiler, leaving it up to cmake to determine which actual compiler to use

The latter would be excess generality unless there are other uses cases besides the situation where someone wants to compiler C code with a C++ compiler. Are there any such use cases? If not, a boolean `force-cpp` property (only supported by the C target) would be sufficient. I agree with @cmnrd that this would be much cleaner than the string matching performed in this code.

Reply:
I think we are all in agreement here about this solution being not ideal.
I don't think `force-cpp` is any better than `c++-mode`, since it suffers from the very same ambiguities. At least `use-cpp-compiler` is very forthcoming about the intent behind the flag.

To summarize the issue:
A C++ compiler cannot be used on `.c` files in CMake. Therefore, `compiler: g++` with `cmake: true` is a compile error. Rather than this error, we switch internally to a `cpp-mode` (only in the CMake-based compiler) and then use `g++`, avoiding this error. This makes the behavior pretty consistent with the scenario where `compiler: g++` and `cmake: false`. To an end-user, both of these scenarios have the same outcome.

Reply:
Maybe we can leave this problem to be solved by the user. If there is a specific line that needs to be added when compiling C files with cmake using a C++ compiler, then couldn't we add this line via `cmake-include`?

Reply:
Yes, that could work for user-provided `.c` files. However, the user would also have to know the name of our generated `.c`and platform files. This isn't so simple when federates are involved, since each federate will have its own unique `.c` file with a unique name, and a single `cmake-include` file cannot cover all these scenarios elegantly I think.

Reply:
Hmm, I see. We could introduce a variable `LF_MAIN_SOURCES` that is a list of all source files and is available to the user. But I am not sure if this is more 'elegant'.

I would also be on board with the `use-cpp-compiler` property that you suggest above.

Reply:
I suppose we all agree on the `use-cpp-compiler` property then?

Reply:
@edwardalee pointed out that we could use the `CCpp` target to activate this mode. I think this could potentially be the best solution. I also very much like @cmnrd's `LF_MAIN_SOURCES` idea. 

I think it would be better if we can avoid adding a new target property. To me, the following program seems to have redundant information:
```
target C {
    cmake: true,
    compiler: g++,
    use-cpp-compiler: true
```

Adding this target property also means that the following two programs will yield different outcomes:
```
target C {
    cmake: true,    // CMake is enabled.
    compiler: g++.  // Compile error.
```
and
```
target C {
    cmake: false,   // CMake is disabled.
    compiler: g++   // No compile error.
```

Reply:
Is the `CCpp` target still a thing? If we plan on using it, there should be tests for it. But considering those tests would be exact copies of the C tests, I don't think this would make much sense.

It has been pointed out before the the `compiler` directive is an awkward combination with `cmake`, because cmake is supposed to abstract away which compiler should be used.

I think the corner case we're trying to address here is that we want a way to specify that we 1) want to use cmake and 2) want to force cmake to use a c++ compiler on C files. Then, why not have a target property called `cmake-force-cpp` (or something similar) that clearly only makes sense if `cmake: true`.

Reply:
i think the idea here is that we repurpose the CCpp target to simply designate C/C++ combos, and that it will actually be the C target with minor adjustments to make work with C++ code. It would not need many tests, just some that combine C and C++.

A meta issue: I feel we are stuck on this issue, and in the spirit of a doacracy, I think Soroush needs to make the decision and move on. We have all provided valuable input, and as near as I can see, there is no perfect answer.

Reply:
I don't want to block progress, but I don't think the CCpp approach will not generally work. What should happen when I write a main reactor in a file that declares `target C` but I imports a reactor Foo in a file that declares `target CCpp`?

What's the downside of `cmake-force-cpp`? I really think it offers more flexibility.

Reply:
> What should happen when I write a main reactor in a file that declares target C but I imports a reactor Foo in a file that declares target CCpp?

I think in this case, we can treat it like any other target mismatch and reject it (we can revisit this in the future for federates). The fix would be to change the target of the file with the main reactor to `CCpp`.

Reply:
But it is _not_ a target mismatch; both are written in C. It is a build configuration issue.

Reply:
I'm not sure understand. A file that declares CCpp is a mix of C and C++, not just C. Seems like a target mismatch to me. Like a C target file importing a TypeScript reactor.

Reply:
`C` and `TypeScript` use a different runtime. `C` and `CCpp` use the same runtime. The only difference between the two is the way they are built (using a C compiler vs. a C++ compiler). This is why I think the issue is best described as a build configuration issue, not a target mismatch. Before cmake came into the picture, this distinction could already be made using the `compiler` property. The problem is that this trick is not a good match with cmake. I think it's a bit odd to (re)instate a dedicated target just to deal with a particular corner case in a particular build system. Not long ago @Soroosh129 had suggested to decommission the `CCpp` target, which I thought was a good idea.